### PR TITLE
Add amp-ad support for AppVador

### PR DESCRIPTION
**Pull request by adg-dev**
We added "AppVador" that the type of amp-ad.
Would you merge this commit?

We already signed a CLA.

CLA name: adg-dev
Corporation Name: Supership Inc.

### CODE REVIEW

File: ads/appvador.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
```
Comment (zhouyx) [inline ads/appvador.js]: nit: 2017

File: ads/appvador.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+ */
+
+import {writeScript, validateData} from '../3p/3p';
+
+/**
+ * @param {!Window} global
+ * @param {!Object} data
+ */
+export function appvador(global, data) {
+  validateData(data, ['id'], ['options','jsType','customScript']);
```
Comment (zhouyx) [inline ads/appvador.js]: nit: space after comma

File: ads/appvador.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+  const container = global.document.getElementById('c');
+  const apvDiv = global.document.createElement('div');
+  apvDiv.setAttribute('id','apvad-' + data.id);
```
Comment (zhouyx) [inline ads/appvador.js]: nit: space after comma

File: ads/appvador.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  container.appendChild(apvDiv);
+
+  const scriptUrl = data.customScript ? data.customScript :
```
Comment (zhouyx) [inline ads/appvador.js]: nit: use `#encodeURIComponent`

File: ads/appvador.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+      'https://cdn.apvdr.com/js/' +
+      (data.jsType ? data.jsType : 'VastAdUnit') + '.min.js';
```
Comment (zhouyx) [inline ads/appvador.js]: encodeURIComponent. same below

File: ads/appvador.md [hunk 6]
```diff
@@ -0,0 +1,37 @@
+<!---
+Copyright 2016 The AMP HTML Authors. All Rights Reserved.
```
Comment (zhouyx) [inline ads/appvador.md]: nit: 2017

File: ads/appvador.md [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    type="appvador"
+    data-id="7393d415ebe7366b0cc24478c5675ee0">
+</amp-ad>
+```
+
+## Configuration
+
+For semantics of configuration, please contact [AppVador](http://www.appvador.com/).
+
+Supported parameters:
```
Comment (zhouyx) [inline ads/appvador.md]: please indicate that `data-id` is required, while others are optional.

File: validator/testdata/feature_tests/ads.html [hunk 8]
```diff
@@ -435,6 +436,12 @@ <h2>AppNexus with JSON based configuration multi ad</h2>
       json='{"pageOpts":{"member": 958}, "adUnits": [{"disablePsa": true, "tagId": 6063968,"sizes": [300,250],"targetId": "apn_ad_1"}, {"tagId": 6063968,"sizes": [728,90],"targetId":"apn_ad_2"}]}'>
   </amp-ad>
 
+  <h2>AppVador</h2>
```
Comment (zhouyx) [inline validator/testdata/feature_tests/ads.html]: There is no need to update this file.

File: ads/_config.js [hunk 9]
```diff
@@ -185,6 +185,16 @@ export const adConfig = {
     renderStartImplemented: true,
   },
 
+  appvador: {
+    prefetch: [
```
Comment (zhouyx) [inline ads/_config.js]: We provide a `renderStart` and `noContent` API that help to provide better UX. We highly recommend implementing them. 
Please refer to https://github.com/ampproject/amphtml/blob/master/ads/README.md#available-apis and let me know if there's any question. Thank you.

File: examples/ads.amp.html [hunk 10]
```diff
@@ -457,6 +458,12 @@ <h2>AppNexus with JSON based configuration multi ad</h2>
       json='{"pageOpts":{"member": 958}, "adUnits": [{"disablePsa": true, "tagId": 6063968,"sizes": [300,250],"targetId": "apn_ad_1"}, {"tagId": 6063968,"sizes": [728,90],"targetId":"apn_ad_2"}]}'>
   </amp-ad>
 
+  <h2>AppVador</h2>
```
Comment (zhouyx) [inline examples/ads.amp.html]: I see "[APVAD] VAST Response is empty." when testing with this example, and I see a blank rect with text "filler tag". Wondering would it be more appropriate to use AMP provided `noContent` API in this case? 
Also we generally suggest adding rendering ad example (a fake one would be great) for testing purpose. Thank you.

**Comment by lannka** [comment]
@adg-dev please address the comments and let me know when you're ready for the 2nd round :-)

**Comment by s-kageyama** [comment]
Thank you for reviewing.
Sorry for late reply
I fixed some issues that were pointed out.

File: ads/_config.js (hunk 9 above)
Comment (lannka) [reply ads/_config.js]: @adg-dev did you get into this?

**Comment by lannka** [comment]
please also address this comment: https://github.com/ampproject/amphtml/pull/12163#discussion_r157035906

File: ads/_config.js (hunk 9 above)
Comment (adg-dev) [reply ads/_config.js]: @lannka add this on last commit. please check again.

File: ads/appvador.js (hunk 4 above)
Comment (adg-dev) [reply ads/appvador.js]: 'data.customScript' is not a part of url. so cannot use #encodeURIComponent

File: examples/ads.amp.html (hunk 10 above)
Comment (adg-dev) [reply examples/ads.amp.html]: I replaced a fake one. Please check again.
Thank you.

**Comment by adg-dev** [comment]
@lannka 
I fixed about `renderStart`.
Please check again.

File: ads/appvador.js [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+ */
+export function appvador(global, data) {
+  validateData(data, ['id'], ['options', 'jsType', 'customScriptSrc']);
+
+  const container = global.document.getElementById('c');
+  const apvDiv = global.document.createElement('div');
+  apvDiv.setAttribute('id', 'apvad-' + data.id);
+  container.appendChild(apvDiv);
+
+  const scriptUrl = data.customScriptSrc ? data.customScriptSrc :
```
Comment (zhouyx) [inline ads/appvador.js]: I would still suggest using `encodeURI(data.customScriptSrc)` for safety. But up to you, let me know.

File: ads/appvador.js [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+      'https://cdn.apvdr.com/js/' +
+      (data.jsType ? encodeURIComponent(data.jsType) : 'VastAdUnit') + '.min.js';
```
Comment (zhouyx) [inline ads/appvador.js]: nit: exceeds 80 chars

File: ads/appvador.js [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+export function appvador(global, data) {
+  validateData(data, ['id'], ['options', 'jsType', 'customScriptSrc']);
+
+  const container = global.document.getElementById('c');
+  const apvDiv = global.document.createElement('div');
+  apvDiv.setAttribute('id', 'apvad-' + data.id);
+  container.appendChild(apvDiv);
+
+  const scriptUrl = data.customScriptSrc ? data.customScriptSrc :
+      'https://cdn.apvdr.com/js/' +
```
Comment (zhouyx) [inline ads/appvador.js]: nit: indentation

File: ads/appvador.js (hunk 12 above)
Comment (adg-dev) [reply ads/appvador.js]: I fixed it.

File: ads/appvador.js (hunk 13 above)
Comment (adg-dev) [reply ads/appvador.js]: I fixed it.

File: ads/appvador.js (hunk 11 above)
Comment (adg-dev) [reply ads/appvador.js]: I need to substitute test script url for `date.customScriptSrc`.
So, I can not use `encodeURI`.

**Comment by adg-dev** [comment]
@zhouyx 
thank you for your checking.
I fixed issues. Please review again.

**Comment by rsimha-amp** [comment]
@lannka @adg-dev This PR is causing lint errors on `master` because it was tested a week ago, and `gulp lint` has become more strict since then.

**Comment by rsimha-amp** [comment]
See https://travis-ci.org/ampproject/amphtml/jobs/339682955

**Pull request merged**